##**Домашнее задание по теме "Поиск аномалий в данных и работа с пропущенными значениями"**

**Задание:**

Будем осуществлять работу с непростым набором данных “horse_data.csv” о состоянии здоровья лошадей, испытывающих кишечные колики. Цель – максимально корректно заполнить пропуски.

Задание 1. Загрузка данных
Изучить представленный набор данных на основе описания его столбцов в файле “horse_data.names” , загрузить его и оставить 8 столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.

Задание 2. Первичное изучение данных
Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.

Примечание:
Домашнее задание сдается ссылкой Google Colab. Мы не сможем проверить его или помочь, если вы пришлете:

файлы;
архивы;
скриншоты кода.
Все обсуждения и консультации по выполнению домашнего задания ведутся только на соответствующем канале в Discord.

Как правильно задавать вопросы аспирантам, преподавателям и коллегам:

Прежде чем задать вопрос, попробуйте найти ответ в интернете. Навык самостоятельного поиска информации — один из важнейших. Каждый практикующий специалист любого уровня делает это ежедневно.

Сформулируйте вопрос по алгоритму:

Что я делаю?
Какого результата я ожидаю?
Как фактический результат отличается от ожидаемого?
Что я уже попробовал сделать, чтобы исправить проблему?
По возможности прикрепите к вопросу скриншоты либо ссылки на код. Не выкладывайте все решение, оставляйте только проблемный и воспроизводимый участок кода

**Решение:**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib as plt


Задание 1. Загрузка данных


Нужно изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения.

In [ ]:
names =['surgery?', 'Age', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 'pain', 'outcome']
df = pd.read_csv('horse_data.csv', usecols =[0,1,3,4,5,6,10,22],names = names, na_values ='?',sep = ',')


1 столбец - 'surgery?' - операция? Значения: 1 - лошадь была прооперирована; 2 - лечение проводилось без операции.

2 столбец: Age: 1 = Adult horse (взрослая лошадь); 2 = Young (< 6 months) (молодая лошадь).

3 столбец: 'rectal temperature' - температура.Нормальная температура составляет 37,8. Измеряется в градусах по Цельсию. Повышенная - может быть при инфекции. Пониженная - при отложенном шоковом состоянии.

4 столбец - 'pulse' частота сердечных сокращений в ударах в минуту.30-40 - для взрослых животных. У спортивных лошадей - 20-25, у больных  - может быть более 40.

5 столбец - 'respiratory rate' - частота дыхания. В норме составляет от 8 - до 10.

6 столбец - 'temperature of extremities' - субъективный показатель о температуре конечностей. Значения: 1 = Нормальный, 2 = Теплый, 3 = Прохладный, 4 = Холодный.

7 столбец - 'pain' - субъективное суждение об уровне боли лошади. Значения: 1 = отсутствие боли (сигнал тревоги!), 2 = подавленный. Не следует рассматривать как упорядоченную или дискретную переменную!

8 столбец - 'Outcome' - Итог, значения: 1 = жив; 2 = умер; 3 = был подвергнут эвтаназии.

In [ ]:
df.head()

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0       2.0    1                38.5   66.0              28.0   
1       1.0    1                39.2   88.0              20.0   
2       2.0    1                38.3   40.0              24.0   
3       1.0    9                39.1  164.0              84.0   
4       2.0    1                37.3  104.0              35.0   

   temperature of extremities  pain  outcome  
0                         3.0   5.0      2.0  
1                         NaN   3.0      3.0  
2                         1.0   3.0      1.0  
3                         4.0   2.0      2.0  
4                         NaN   NaN      2.0

Задание 2. Первичное изучение данных

In [ ]:
df.describe()


surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

Проверим соответствие внесенных данных в столбцах строго определенному списку возможных значений. Такой список значений установлен для: 'surgery?', 'age', 'temperature_of_extremities', 'pain', 'outcome'. Однако 'pain', как указано в описании, не следует рассматривать, как упорядоченную дискретную величину. Поэтому ее исключим.

In [ ]:
value_counts = ['surgery?', 'Age', 'temperature of extremities', 'outcome']
for value in value_counts:
    print(df[value].value_counts())

surgery?
1.0    180
2.0    119
Name: count, dtype: int64
Age
1    276
9     24
Name: count, dtype: int64
temperature of extremities
3.0    109
1.0     78
2.0     30
4.0     27
Name: count, dtype: int64
outcome
1.0    178
2.0     77
3.0     44
Name: count, dtype: int64


Анализ полученных значений показывает, что в столбце Age  встречается неустановленное значение "9" на фоне отсутствия предусмотренного - "2". Вероятнее всего была допущена ошибка в данных. Проверим это сравнив данные по выборкам для "9" и "1".

In [ ]:
df[df.Age == 1].mean()

surgery?                       1.410909
Age                            1.000000
rectal temperature            38.124545
pulse                         67.456693
respiratory rate              28.238739
temperature of extremities     2.364035
pain                           2.938596
outcome                        1.552727
dtype: float64

In [ ]:
df[df.Age == 9].mean()

surgery?                        1.250000
Age                             9.000000
rectal temperature             38.645000
pulse                         123.363636
respiratory rate               54.600000
temperature of extremities      2.125000
pain                            3.117647
outcome                         1.541667
dtype: float64

Здесь уже наглядно видно, что в выборке с возрастом "9" по сравнению с выборкой, где данный показатель равен "1": значительно увеличен пульс, частота дыхания, чаще проводились операции и хуже выживаемость. Очевидно, что значение "9" по всем параметрам соответсвует значению "2".
Поэтому произведем замену.

In [ ]:
df.loc[(df.Age == 9), 'Age'] = 2

И проверим результат:

In [ ]:
value_counts = ['Age']
for value in value_counts:
    print(df[value].value_counts())

Age
1    276
2     24
Name: count, dtype: int64


Рассчитаем показатели моды для категориальных величин:

In [ ]:
print(df['surgery?'].mode()[0])
print(df['Age'].mode()[0])
print(df['temperature of extremities'].mode()[0])
print(df['pain'].mode()[0])
print(df['outcome'].mode()[0])

1.0
1
3.0
3.0
1.0


Рассчитаем уникальные значения для категориальных величин:

In [ ]:
print(pd.unique(df['surgery?']))
print(pd.unique(df['Age']))
print(pd.unique(df['temperature of extremities']))
print(pd.unique(df['pain']))
print(pd.unique(df['outcome']))

[ 2.  1. nan]
[1 2]
[ 3. nan  1.  4.  2.]
[ 5.  3.  2. nan  4.  1.]
[ 2.  3.  1. nan]


После исправления столбца Age - все корректно, за исключением значений nan.

Посчитаем базовые статистики для непрерывных случайных величин:

In [ ]:
def stat(column):
    print(f'Среднее значение столбца {column.name} - {column.mean()}')
    print(f'Медианное значение столбца {column.name} - {column.median()}')
    print(f'Модальное значение столбца {column.name} - {column.mode()[0]}')
    print(f'Стандартное отклонение столбца {column.name} - {column.std()}')
    print(f'Дисперсия столбца {column.name} - {column.var()}')

In [ ]:
stat(df['rectal temperature'])

Среднее значение столбца rectal temperature - 38.16791666666667
Медианное значение столбца rectal temperature - 38.2
Модальное значение столбца rectal temperature - 38.0
Стандартное отклонение столбца rectal temperature - 0.7322886641121578
Дисперсия столбца rectal temperature - 0.5362466875871686


In [ ]:
stat(df['respiratory rate'])

Среднее значение столбца respiratory rate - 30.417355371900825
Медианное значение столбца respiratory rate - 24.5
Модальное значение столбца respiratory rate - 20.0
Стандартное отклонение столбца respiratory rate - 17.64223138513467
Дисперсия столбца respiratory rate - 311.2483282466308


In [ ]:
stat(df['pulse'])

Среднее значение столбца pulse - 71.91304347826087
Медианное значение столбца pulse - 64.0
Модальное значение столбца pulse - 48.0
Стандартное отклонение столбца pulse - 28.630556660735
Дисперсия столбца pulse - 819.7087747035573


Посчитаем выбросы через межквартильный размах:

In [ ]:
def find_outlier(df, param):
    q1 = df[param].quantile(0.25)
    q3 = df[param].quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr)
    upper_bound = q3 + (1.5 * iqr)
    df2= df[~df[param].between(lower_bound, upper_bound, inclusive=True)]
    return df2.loc[df2[param].notnull()]

In [ ]:
q1 = df['rectal temperature'].quantile(0.25)
q3 = df['rectal temperature'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['rectal temperature'].between(lower_bound, upper_bound)]
remove_outliers

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0         2.0    1                38.5   66.0              28.0   
1         1.0    1                39.2   88.0              20.0   
2         2.0    1                38.3   40.0              24.0   
3         1.0    2                39.1  164.0              84.0   
4         2.0    1                37.3  104.0              35.0   
..        ...  ...                 ...    ...               ...   
292       1.0    1                37.0   66.0              20.0   
294       2.0    1                38.5   40.0              16.0   
296       2.0    1                37.2   72.0              24.0   
297       1.0    1                37.5   72.0              30.0   
299       1.0    1                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           NaN   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           NaN   NaN      2.0  
..                          ...   ...      ...  
292                         1.0   4.0      2.0  
294                         1.0   2.0      1.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
299                         NaN   NaN      3.0  

[226 rows x 8 columns]

При норме данного показателя - 37.8. в выбросы попали как высокие, так и низкие значения.Высокие значения, согласно описанию данных, могут быть связаны с инфекцией, а низкие - с последними стадиями шока.

In [ ]:
q1 = df['respiratory rate'].quantile(0.25)
q3 = df['respiratory rate'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['respiratory rate'].between(lower_bound, upper_bound)]
remove_outliers

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0         2.0    1                38.5   66.0              28.0   
1         1.0    1                39.2   88.0              20.0   
2         2.0    1                38.3   40.0              24.0   
4         2.0    1                37.3  104.0              35.0   
6         1.0    1                37.9   48.0              16.0   
..        ...  ...                 ...    ...               ...   
294       2.0    1                38.5   40.0              16.0   
296       2.0    1                37.2   72.0              24.0   
297       1.0    1                37.5   72.0              30.0   
298       1.0    1                36.5  100.0              24.0   
299       1.0    1                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           NaN   3.0      3.0  
2                           1.0   3.0      1.0  
4                           NaN   NaN      2.0  
6                           1.0   3.0      1.0  
..                          ...   ...      ...  
294                         1.0   2.0      1.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         NaN   NaN      3.0  

[225 rows x 8 columns]

В выборку попали лошади с очень высокими значениями частоты дыхания, которые значительно превышают норму.

In [ ]:
q1 = df['pulse'].quantile(0.25)
q3 = df['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['pulse'].between(lower_bound, upper_bound)]
remove_outliers

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0         2.0    1                38.5   66.0              28.0   
1         1.0    1                39.2   88.0              20.0   
2         2.0    1                38.3   40.0              24.0   
4         2.0    1                37.3  104.0              35.0   
6         1.0    1                37.9   48.0              16.0   
..        ...  ...                 ...    ...               ...   
295       1.0    1                 NaN  120.0              70.0   
296       2.0    1                37.2   72.0              24.0   
297       1.0    1                37.5   72.0              30.0   
298       1.0    1                36.5  100.0              24.0   
299       1.0    1                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           NaN   3.0      3.0  
2                           1.0   3.0      1.0  
4                           NaN   NaN      2.0  
6                           1.0   3.0      1.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         NaN   NaN      3.0  

[271 rows x 8 columns]

В выбросы попали очень высокие значения пульса в кластере взрослых лошадей, которые могут быть вызваны болезненными поражениями или цикличным шоком. Этот факт косвенно подтверждают соответсвенно высокие значения
respiratory_rate, rectal temperature и pain.
Из 5 лошадей выжила только та, что была прооперирована. Следовательно, можно сделать вывод: если у взрослой лошади значительно повышены данные показатели (в выбросах по значению), то ей необходима операция, иначе шансов на выживаемость нет.

Задание 3.  Работа с пропусками

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         300 non-null    int64  
 2   rectal temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory rate            242 non-null    float64
 5   temperature of extremities  244 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


Рассчитаем процент порпущенных данных по категориям:

In [ ]:
(df.isna().mean() * 100).round(2)

surgery?                       0.33
Age                            0.00
rectal temperature            20.00
pulse                          8.00
respiratory rate              19.33
temperature of extremities    18.67
pain                          18.33
outcome                        0.33
dtype: float64

Создадим копию df, в которой заполним пропуски:

In [ ]:
df_new = df.copy()
df_new


surgery?  Age  rectal temperature  pulse  respiratory rate  \
0         2.0    1                38.5   66.0              28.0   
1         1.0    1                39.2   88.0              20.0   
2         2.0    1                38.3   40.0              24.0   
3         1.0    2                39.1  164.0              84.0   
4         2.0    1                37.3  104.0              35.0   
..        ...  ...                 ...    ...               ...   
295       1.0    1                 NaN  120.0              70.0   
296       2.0    1                37.2   72.0              24.0   
297       1.0    1                37.5   72.0              30.0   
298       1.0    1                36.5  100.0              24.0   
299       1.0    1                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           NaN   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           NaN   NaN      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         NaN   NaN      3.0  

[300 rows x 8 columns]

Заполним пропуски модой для категориальных атрибутов, в который есть пропуски:

In [ ]:
df_new['surgery?'] = df_new['surgery?'].fillna(df['surgery?'].mode()[0])
df_new['temperature of extremities'] = df_new['temperature of extremities'].fillna(df['temperature of extremities'].mode()[0])
df_new['pain'] = df_new['pain'].fillna(df['pain'].mode()[0])
df_new['outcome'] = df_new['outcome'].fillna(df['outcome'].mode()[0])
df_new

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0         2.0    1                38.5   66.0              28.0   
1         1.0    1                39.2   88.0              20.0   
2         2.0    1                38.3   40.0              24.0   
3         1.0    2                39.1  164.0              84.0   
4         2.0    1                37.3  104.0              35.0   
..        ...  ...                 ...    ...               ...   
295       1.0    1                38.2  120.0              70.0   
296       2.0    1                37.2   72.0              24.0   
297       1.0    1                37.5   72.0              30.0   
298       1.0    1                36.5  100.0              24.0   
299       1.0    1                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           3.0   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           3.0   3.0      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         3.0   3.0      3.0  

[300 rows x 8 columns]

Запоним средним значением там, где присутствуют выбросы и пропуски:

In [ ]:
df_new[['rectal temperature', 'pulse', 'respiratory rate']]= df_new[['rectal temperature', 'pulse', 'respiratory rate']].fillna(df[['rectal temperature', 'pulse', 'respiratory rate']].mean()).round(1)
df_new

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0         2.0    1                38.5   66.0              28.0   
1         1.0    1                39.2   88.0              20.0   
2         2.0    1                38.3   40.0              24.0   
3         1.0    2                39.1  164.0              84.0   
4         2.0    1                37.3  104.0              35.0   
..        ...  ...                 ...    ...               ...   
295       1.0    1                38.2  120.0              70.0   
296       2.0    1                37.2   72.0              24.0   
297       1.0    1                37.5   72.0              30.0   
298       1.0    1                36.5  100.0              24.0   
299       1.0    1                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           3.0   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           3.0   3.0      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         3.0   3.0      3.0  

[300 rows x 8 columns]

Проверим количество пропусков:

In [ ]:
pd.isna(df_new).sum()

surgery?                      0
Age                           0
rectal temperature            0
pulse                         0
respiratory rate              0
temperature of extremities    0
pain                          0
outcome                       0
dtype: int64

Сравним статистические показатели до и после внесения изменений.

In [ ]:
df.describe()

surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.080000           38.167917   71.913043   
std      0.490305    0.271746            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

In [ ]:
df_new.describe()

surgery?         Age  rectal temperature       pulse  \
count  300.000000  300.000000          300.000000  300.000000   
mean     1.396667    1.080000           38.174333   71.912000   
std      0.490023    0.271746            0.654831   27.457472   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.900000   48.000000   
50%      1.000000    1.000000           38.200000   68.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        300.000000                  300.000000  300.000000  300.000000  
mean          30.414000                    2.470000    2.960000    1.550000  
std           15.838953                    0.975845    1.181689    0.736642  
min            8.000000                    1.000000    1.000000    1.000000  
25%           20.000000                    1.000000    2.000000    1.000000  
50%           30.000000                    3.000000    3.000000    1.000000  
75%           34.250000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000

Сравним Моду:

In [ ]:
df.mode()

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0       1.0    1                38.0   48.0              20.0   

   temperature of extremities  pain  outcome  
0                         3.0   3.0      1.0

In [ ]:
df_new.mode()

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0       1.0    1                38.2   48.0              30.4   

   temperature of extremities  pain  outcome  
0                         3.0   3.0      1.0

Сравним Медиану:

In [ ]:
df.median()

surgery?                       1.0
Age                            1.0
rectal temperature            38.2
pulse                         64.0
respiratory rate              24.5
temperature of extremities     3.0
pain                           3.0
outcome                        1.0
dtype: float64

In [ ]:
df_new.median()

surgery?                       1.0
Age                            1.0
rectal temperature            38.2
pulse                         68.0
respiratory rate              30.0
temperature of extremities     3.0
pain                           3.0
outcome                        1.0
dtype: float64

Мы наблюдаем незначительные отличия по показателям: rectal temperature, pulse, respiratory rate - в них присутствовали выбросы,
поэтому подсчет осуществляли корректным с помощью среднего значения. Отличие может быть
из-за возраста лошади и ее состояния здоровья.

Сравним межквартильный размах:

In [ ]:
Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)
IQR_new = Q3 - Q1
IQR_new

surgery?                       1.0
Age                            0.0
rectal temperature             0.7
pulse                         40.0
respiratory rate              17.5
temperature of extremities     2.0
pain                           2.0
outcome                        1.0
dtype: float64

In [ ]:
Q1_new = df_new.quantile(0.25)
Q3_new = df_new.quantile(0.75)
IQR_new = Q3_new - Q1_new
IQR_new

surgery?                       1.00
Age                            0.00
rectal temperature             0.60
pulse                         40.00
respiratory rate              14.25
temperature of extremities     2.00
pain                           2.00
outcome                        1.00
dtype: float64

Мы наблюдаем незначительные изменения в показателях: respiratory rate и rectal temperature. По ним было максимальное количество пропусков данных. Вероятнее всего обусловлены возрастом лошади.